README
=== README for Geo-EOP version 0.11
=   Generated on Wed Jan 21 12:18:35 2009 by OODoc 1.05

There are various ways to install this module:

 (1) if you have a command-line, you can do:
       perl -MCPAN -e 'install <any package from this distribution>'

 (2) if you use Windows, have a look at http://ppm.activestate.com/

 (3) if you have downloaded this module manually (as root/administrator)
       gzip -d Geo-EOP-0.11.tar.gz
       tar -xf Geo-EOP-0.11.tar
       cd Geo-EOP-0.11
       perl Makefile.PL
       make          # optional
       make test     # optional
       make install

For usage, see the included manual-pages or
    http://search.cpan.org/dist/Geo-EOP-0.11/

Please report problems to
    http://rt.cpan.org/Dist/Display.html?Queue=Geo-EOP



ChangeLog

======== version history of Geo::EOP

version 0.13: Thu Jan 29 16:47:24 CET 2009

	Improvements:

	- add Geo::EOP::Landsat, which converts some Landsat product
	  meta-data into the EOP structure.

version 0.12: Wed Jan 28 11:10:21 CET 2009

	Changes:

	- remove 'READER'|'WRITER'|'RW' first parameter from new()
	  in Geo::EOP and Geo::EOP::Envisat.

	Fixes:

	- do not use /bin/pwd in t/99pod.t

	Improvements:

	- add Geo::EOP::from()

	- included examples/read_eop.pl

version 0.11: Wed Jan 21 12:43:57 CET 2009

	Changes:

	- Geo::EOP is now fully based on a Geo::GML object.

	- Geo::GML uses prefixed keys, so Geo::EOP uses prefixed keys as
	  well.

	- version() and new(version) renamed to resp. eopVersion() and
	  new(eop_version) to avoid clash with GML version.

	- remove NS_GML* definitions from Geo::EOP::Util: get them
	  yourself from Geo::GML::Util.

	Improvements:

	- added eop spec 1.2.1 and tests in t/22eop121.t

	- added Geo::EOP::Envisat as demonstration implementation.  See
	  also examples/envisat.pl

	- use Geo::GML's version of xlink1.0.0

	- move schematron info from examples (which they aren't) into
	  a seperate directory.  Schematron information is not (yet)
	  supported.

version 0.10: Tue Jul  8 13:49:23 CEST 2008

	- initial release


Makefile.PL
use ExtUtils::MakeMaker;

use 5.008;

WriteMakefile
 ( NAME	 	=> 'Geo::EOP'
 , VERSION	=> '0.13'
 , PREREQ_PM    =>
   { Geo::GML            => 0.13
   , XML::Compile        => '1.00'
   , XML::Compile::Cache => '0.90'

   , Test::More    => 0.54
   , Test::Pod     => '1.00'
   }

 , EXE_FILES    => []
 , AUTHOR       => 'Mark Overmeer'
 , ABSTRACT     => 'Earth Observation Products meta-data'
 , LICENSE      => 'perl'
 );

### used by oodist during production of distribution
sub MY::postamble { <<'__POSTAMBLE' }

# for DIST
RAWDIR          = ../public_html/geo-eop/raw
DISTDIR         = ../public_html/geo-eop/source
LICENSE         = artistic
EXTENDS		= ../XMLCompile:../XMLCache:../GeoGML

# for POD
FIRST_YEAR      = 2008
EMAIL           = perl@overmeer.net
WEBSITE         = http://perl.overmeer.net/geo/
PODTAIL		= ../Geo-ogc-podtail
__POSTAMBLE


lib/Geo/EOP/Util.pm
# Copyrights 2008-2009 by Mark Overmeer.
#  For other contributors see ChangeLog.
# See the manual pages for details on the licensing terms.
# Pod stripped from pm file by OODoc 1.06.
use warnings;
use strict;

package Geo::EOP::Util;
use vars '$VERSION';
$VERSION = '0.13';

use base 'Exporter';

our @hmas      = qw/NS_ATM_ESA NS_OHR_ESA NS_SAR_ESA NS_HMA_ESA/;
our @eops      = qw/NS_ATM_ESA NS_OPT_ESA NS_SAR_ESA NS_EOP_ESA/;

our @EXPORT    = (@hmas, @eops);

our @EXPORT_TAGS =
 ( hma10     => \@hmas
 , eop11     => \@eops
 , eop12beta => \@eops
 , eop121    => \@eops
 );


use constant NS_ATM_ESA => 'http://earth.esa.int/atm';
use constant NS_SAR_ESA => 'http://earth.esa.int/sar';

# HMA before renaming
use constant NS_OHR_ESA => 'http://earth.esa.int/ohr';
use constant NS_HMA_ESA => 'http://earth.esa.int/hma';

# EOP renaming
use constant NS_OPT_ESA => 'http://earth.esa.int/opt';
use constant NS_EOP_ESA => 'http://earth.esa.int/eop';

1;


lib/Geo/EOP.pm
# Copyrights 2008-2009 by Mark Overmeer.
#  For other contributors see ChangeLog.
# See the manual pages for details on the licensing terms.
# Pod stripped from pm file by OODoc 1.06.
use warnings;
use strict;

package Geo::EOP;
use vars '$VERSION';
$VERSION = '0.13';

use base 'Geo::GML';

use Geo::EOP::Util;   # all
use Geo::GML::Util  qw/:gml311/;

use Log::Report 'geo-eop', syntax => 'SHORT';
use XML::Compile::Util  qw/unpack_type pack_type type_of_node/;
use Math::Trig          qw/rad2deg deg2rad/;

# map namespace always to the newest implementation of the protocol
my %ns2version =
  ( &NS_HMA_ESA => '1.0'
  , &NS_EOP_ESA => '1.2.1'
  );

# list all available versions
# It is a pity that not all schema use the same prefixes... sometimes,
# the dafault prefix is used... therefore, we have to configure all that
# manually.

my @stdprefs =   # will be different in the future
 ( sar => NS_SAR_ESA
 , atm => NS_ATM_ESA
 , gml => NS_GML_311
 );

my %info =
  ( '1.0'     =>
      { prefixes    => {hma => NS_HMA_ESA, ohr => NS_OHR_ESA, @stdprefs}
      , eop_schemas => [ 'hma1.0/{eop,sar,opt,atm}.xsd' ]
      , gml_schemas => [ 'eop1.1/gmlSubset.xsd' ]
      , gml_version => '3.1.1eop'
      }

  , '1.1'     =>
      { prefixes    => {eop => NS_EOP_ESA, opt => NS_OPT_ESA, @stdprefs}
      , eop_schemas => [ 'eop1.1/{eop,sar,opt,atm}.xsd' ]
      , gml_schemas => [ 'eop1.1/gmlSubset.xsd' ]
      , gml_version => '3.1.1eop'
      }

  , '1.2beta' =>
      { prefixes    => {eop => NS_EOP_ESA, opt => NS_OPT_ESA, @stdprefs}
      , eop_schemas => [ 'eop1.2beta/{eop,sar,opt,atm}.xsd' ]
      , gml_schemas => [ 'eop1.1/gmlSubset.xsd' ]
      , gml_version => '3.1.1eop'
      }

  , '1.2.1' =>
      { prefixes    => {eop => NS_EOP_ESA, opt => NS_OPT_ESA, @stdprefs}
      , eop_schemas => [ 'eop1.2.1/{eop,sar,opt,atm}.xsd' ]
      , gml_schemas => [ 'eop1.2.1/gmlSubset.xsd' ]
      , gml_version => '3.1.1eop'
      }

# , '2.0' =>
#     { eop_schemas => [ 'eop2.0/*.xsd' ]
#     , gml_version => '3.2.1'
#     }

  );

my %measure =
  ( rad_deg   => sub { rad2deg $_[0] }
  , deg_rad   => sub { deg2rad $_[0] }
  , '%_float' => sub { $_[0] / 100 }
  , 'float_%' => sub { sprintf "%.2f", $_[0] / 100 }
  );
sub _convert_measure($@);

# This list must be extended, but I do not know what people need.
my @declare_always = ();


sub new($@) { my $class = shift; $class->SUPER::new('RW', @_) }

sub init($)
{   my ($self, $args) = @_;
    $args->{allow_undeclared} = 1
        unless exists $args->{allow_undeclared};

    my $version  =  $args->{eop_version}
        or error __x"EOP object requires an explicit eop_version";

    unless(exists $info{$version})
    {   exists $ns2version{$version}
            or error __x"EOP version {v} not recognized", v => $version;
        $version = $ns2version{$version};
    }
    $self->{GE_version} = $version;    
    my $info            = $info{$version};

    $args->{version}    = $info->{gml_version};
    if($info->{gml_schemas})  # using own GML 3.1.1 subset
    {   $self->_register_gml_version($info->{gml_version} => {});
    }

    $self->SUPER::init($args);

    $self->prefixes($info->{prefixes});

    (my $xsd = __FILE__) =~ s!\.pm!/xsd!;
    my @xsds    = map {glob "$xsd/$_"}
       @{$info->{eop_schemas} || []}, @{$info->{gml_schemas} || []};

    $self->importDefinitions(\@xsds);

    my $units = delete $args->{units};
    if($units)
    {   if(my $a = $units->{angle})
        {   $self->addHook(type => 'gml:AngleType'
              , after => sub { _convert_measure $a, @_} );
        }
        if(my $d = $units->{distance})
        {   $self->addHook(type => 'gml:MeasureType'
              , after => sub { _convert_measure $d, @_} );
        }
        if(my $p = $units->{percentage})
        {   $self->addHook(path => qr/Percentage/
              , after => sub { _convert_measure $p, @_} );
        }
    }

    $self;
}

sub declare(@)
{   my $self = shift;

    my $direction = $self->direction;

    $self->declare($direction, $_)
        for @_, @declare_always;

    $self;
}


sub from($@)
{   my ($class, $data, %args) = @_;
    my $xml = XML::Compile->dataToXML($data);

    my $product = type_of_node $xml;
    my $version = $xml->getAttribute('version');
    defined $version
        or error __x"no version attribute in root element";

    exists $info{$version}
        or error __x"EOP version {version} not (yet) supported.  Upgrade Geo::EOP or inform author"
             , version => $version;

    my $self    = $class->new(eop_version => $version, %args);
    my $r       = $self->reader($product);
    defined $r
        or error __x"do not understand root node {type}", type => $product;

    ($product, $r->($xml));
}

#---------------------------------


sub eopVersion() {shift->{GE_version}}

#--------------


sub printIndex(@)
{   my $self = shift;
    my $fh   = @_ % 2 ? shift : select;
    $self->SUPER::printIndex($fh
      , kinds => 'element', list_abstract => 0, @_); 
}

# This code will probaby move to Geo::GML
sub _convert_measure($@)   # not $$$$ for right context
{   my ($to, $node, $data, $path) = @_;
    ref $data eq 'HASH'  or return $data;
    my ($val, $from) = @$data{'_', 'uom'};
    defined $val && $from or return $data;

    return $val if $from eq $to;
    my $code = $measure{$from.'_'.$to} or return $data;
    $code->($val);
}

#----------------------


1;
